File: LeetCode/Medium/1190-reverse-substrings-between-each-pair-of-parentheses/1190-reverse-substrings-between-each-pair-of-parentheses.py
class Solution:
    def reverseParentheses(self, s: str) -> str:
        ret = []
        stack = []
        idx = 0
        
        def rev(l, r):
            print(l, r)
            while l < r:
                ret[l], ret[r] = ret[r], ret[l]
                l += 1
                r -= 1
        
        for i in range(len(s)):
            if s[i] == "(":
                stack.append(idx)
            elif s[i] == ")":
                rev(stack.pop(), len(ret) - 1)
            else:
                ret.append(s[i])
                idx += 1
        
        return "".join(ret)

File: LeetCode/Medium/1190-reverse-substrings-between-each-pair-of-parentheses/1190-reverse-substrings-between-each-pair-of-parentheses.go
func reverseParentheses(s string) string {
    ret := make([]byte, 0)
    stack := make([]int, 0)
    idx := 0

    rev := func (l, r int){}
    rev = func (l, r int){
        fmt.Println(l, r)
        for l < r {
            ret[l], ret[r] = ret[r], ret[l]
            l++
            r--
        }
    }

    for i := 0; i < len(s); i++ {
        if s[i] == '(' {
            stack = append(stack, len(ret))
        } else if s[i] == ')' {
            rev(stack[len(stack)-1], len(ret) - 1)
            stack = stack[:len(stack)-1]
        } else {
            ret = append(ret, s[i]) 
            idx++
        }
    }
    
    return string(ret[:])
}